### DISCUSSION TITLE
Custom Parcelers

### DISCUSSION THREAD

**Discussion by gnawf** [arkivanov/Essenty#47] [Ideas]
I have an abstract class that I want to write a custom `Parceler` for per https://developer.android.com/kotlin/parcelize

e.g.

```kotlin
@Parcelize
data class User(val firstName: String, val lastName: String, val age: Int) : Parcelable {
    private companion object : Parceler<User> {
        override fun User.write(parcel: Parcel, flags: Int) {
            // Custom write implementation
        }

        override fun create(parcel: Parcel): User {
            // Custom read implementation
        }
    }
}
```

Cheers.

**Comment by arkivanov** [comment]
Hello! I never tried custom parcelers in multipaltform. But I think expect/actual should work fine.

```kotlin
// commonMain
expect data class User(...) : Parcelable

// androidMain
@Parcelize
actual data class User(...) : Parcelable {
    // Custom Parceler here
}

// Other source sets
actual data class User(...) : Parcelable
```

Please let me know if there are any issues.

**Comment by arkivanov** [comment]
Closing this for now. Feel free to reopen if there are any questions.

**Comment by gnawf** [comment]
Oh sorry, yes your suggestion worked 😄. From memory though, it involved a little bit of boilerplate, so a multiplatform option built into Essenty would be nice.

**Comment by xxfast** [comment]
would be nice if we can support `TypeParceler<ExternalClass, ExternalClassParceler>` in common code, but totally understand if we cant

**Comment by arkivanov** [comment]
@xxfast This would require expect/actual the whole `Parcel` class. Meantime, it's possible to expect/actual custom parcelers in the following way.

```kotlin
// commonMain

expect interface Parceler<T>

@OptIn(ExperimentalMultiplatform::class)
@OptionalExpectation
@Target(AnnotationTarget.TYPE)
expect annotation class WriteWith<P : Parceler<*>>()
```

```kotlin
// androidMain

actual typealias Parceler<T> = kotlinx.parcelize.Parceler<T>

actual typealias WriteWith<P> = kotlinx.parcelize.WriteWith<P>
```

```kotlin
// nonAndroidMain - a shared source source for al targets except Android

actual interface Parceler<T>
```

After that you can expect/actual your custom parcelers as follows.

```kotlin
// commonMain

class ExternalClass(val value: Int)

expect object ExternalClassParceler : Parceler<ExternalClass>

@Parcelize
class MyClass(val external: @WriteWith<ExternalClassParceler> ExternalClass) : Parcelable
```

```kotlin
// androidMain

actual object ExternalClassParceler : Parceler<ExternalClass> {
    override fun create(parcel: Parcel): ExternalClass = ExternalClass(parcel.readInt())

    override fun ExternalClass.write(parcel: Parcel, flags: Int) {
        parcel.writeInt(value)
    }
}
```

```kotlin
// nonAndroidMain

actual object ExternalClassParceler : Parceler<ExternalClass>
```

Yes, there is some boilerplate, but at least it works. It should be also possible to expect/actual `Parcel, with only required methods. In this case `Parceler` would be placed fully in `commonMain` without expect/actual. And so custom parcelers could be placed in `commonMain` as well.

**Comment by arkivanov** [comment]
Converted this issue to a discussion for transparency.

**Comment by xxfast** [comment]
I know this is a couple of months late - but just wanted to say that the [above approach ](https://github.com/arkivanov/Essenty/discussions/47#discussioncomment-2739878)works 👍 thanks @arkivanov 

Just curious - why was `Parceler` and `WriteWith` not included in `com.arkivanov.essenty.parcelable`?

**Reply by arkivanov** [reply]
> Just curious - why was `Parceler` and `WriteWith` not included in `com.arkivanov.essenty.parcelable`?

There are no specific reasons, we can easily add it! I have filed #51. I would gratefully accept a contribution here, in case someone would like to contribute.

**Reply by xxfast** [reply]
@arkivanov Done #52

**Comment by arkivanov** [comment]
Version [0.4.0](https://github.com/arkivanov/Essenty/releases/tag/0.4.0) is released with `Parceler` and `WriteWith` APIs. Thanks everyone for raising this and use cases, and thanks @xxfast for your contribution!
